lib/base/io-engine.cpp

```diff
@@ -16,60 +16,83 @@
+	Shared<AsioConditionVariable>::Ptr cv;
+
+	auto fastPath = [&ie] {
+		auto freeSlots (ie.m_CpuBoundSemaphore.load());
 
-		if (availableSlots < 1) {
-			ioEngine.m_CpuBoundSemaphore.fetch_add(1);
-			IoEngine::YieldCurrentCoroutine(yc);
-			continue;
+		while (freeSlots > 0) {
+			// A failure here is equivalent to loading the latest value into freeSlots
+			if (ie.m_CpuBoundSemaphore.compare_exchange_weak(freeSlots, freeSlots - 1)) {
+				return true;
+			}
 		}
 
-		break;
-	}
-}
+		return false;
+	};
 
-CpuBoundWork::~CpuBoundWork()
-{
-	if (!m_Done) {
-		IoEngine::Get().m_CpuBoundSemaphore.fetch_add(1);
+	while (!fastPath()) {
+		if (!cv) {
+			cv = Shared<AsioConditionVariable>::Make(ie.GetIoContext());
+
+			// The above line may take a little bit, so let's optimistically re-check
+			if (fastPath()) {
+				break;
+			}
+		}
+
+		{
+			std::unique_lock lock (ie.m_CpuBoundWaitingMutex);
+
+			// The above line may take even longer, so let's check again.
+			// Also mitigate lost wake-ups by synchronizing with Done().
+			// On details, see Done() implementation inline comments.
+			if (fastPath()) {
+				break;
+			}
+
+			ie.m_CpuBoundWaiting.emplace_back(strand, cv);
+		}
+
+		cv->Wait(yc);
 	}
 }
 
 void CpuBoundWork::Done()
 {
 	if (!m_Done) {
-		IoEngine::Get().m_CpuBoundSemaphore.fetch_add(1);
-
 		m_Done = true;
-	}
-}
-
-IoBoundWorkSlot::IoBoundWorkSlot(boost::asio::yield_context yc)
-	: yc(yc)
-{
-	IoEngine::Get().m_CpuBoundSemaphore.fetch_add(1);
-}
-
-IoBoundWorkSlot::~IoBoundWorkSlot()
-{
-	auto& ioEngine (IoEngine::Get());
 
-	for (;;) {
-		auto availableSlots (ioEngine.m_CpuBoundSemaphore.fetch_sub(1));
-
-		if (availableSlots < 1) {
-			ioEngine.m_CpuBoundSemaphore.fetch_add(1);
-			IoEngine::YieldCurrentCoroutine(yc);
-			continue;
+		auto& ie (IoEngine::Get());
+
+		// The constructor takes the slow path only if the semaphore is full,
+		// so we only have to wake up constructors if the semaphore was full.
+		// This works because after fetch_add(), fast paths in the constructor will succeed.
+		if (ie.m_CpuBoundSemaphore.fetch_add(1) < 1) {
+			// So now there are only slow path subscribers from just before the fetch_add() to be woken up.
+			// Precisely, only subscribers from just before the fetch_add() which turned 0 to 1.
+
+			decltype(ie.m_CpuBoundWaiting) subscribers;
+
+			{
+				// Due to the mutex, wake-ups obviously may happen only before or after constructor subscriptions.
+				// Ideally, this wake-up happens after any subscription (obviously).
+				std::unique_lock lock (ie.m_CpuBoundWaitingMutex);
+				std::swap(subscribers, ie.m_CpuBoundWaiting);
+			}
+			// But even if it happens before (and gets lost), it's not a problem because now the constructor
+			// will lock the mutex and re-check the semaphore which is already >0 (fast path) due to fetch_add().
+
+			for (auto& [strand, cv] : subscribers) {
+				boost::asio::post(strand, [cv = std::move(cv)] { cv->NotifyOne(); });
+			}
```

Comment:
I'm unclear about what you mean by _But even if it happens before_ in your comments. However, if I'm interpreting this correctly, wouldn't these wake-ups become meaningless if a new coroutine constantly takes the now available free slot from above?

Reply:
Yes, that's the downside of unfair scheduling. 🤷‍♂️